File: virtual/GR-06/codigo/cholesky/cholesky_paralelo.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <time.h>
#include <omp.h>

#define MAX_SIZE 500
#define RUTA_BASE "../../matrices/def_pos_inv/"

// Leer CSV
int leer_csv(const char *ruta_archivo, double A[MAX_SIZE][MAX_SIZE], int *n) {
    FILE *f = fopen(ruta_archivo, "r");
    if (!f) {
        perror("No se pudo abrir el archivo CSV");
        return 0;
    }

    char linea[8192];
    int fila = 0;
    while (fgets(linea, sizeof(linea), f)) {
        int col = 0;
        char *token = strtok(linea, ",");
        while (token && col < MAX_SIZE) {
            A[fila][col++] = atof(token);
            token = strtok(NULL, ",");
        }
        fila++;
        if (fila > MAX_SIZE) break;
    }

    *n = fila;
    fclose(f);
    return 1;
}

// Imprimir matriz
void imprimir(double A[MAX_SIZE][MAX_SIZE], int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            printf("%10.4lf ", A[i][j]);
        }
        printf("\n");
    }
}

// Descomposición de Cholesky
int cholesky(double A[MAX_SIZE][MAX_SIZE], double L[MAX_SIZE][MAX_SIZE], int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j <= i; j++) {
            double suma = 0;
            for (int k = 0; k < j; k++)
                suma += L[i][k] * L[j][k];

            if (i == j) {
                double val = A[i][i] - suma;
                if (val <= 0) return 0;
                L[i][j] = sqrt(val);
            } else {
                L[i][j] = (1.0 / L[j][j]) * (A[i][j] - suma);
            }
        }
    }
    return 1;
}

// Ly = b
void forward_sub(double L[MAX_SIZE][MAX_SIZE], double b[MAX_SIZE], double y[MAX_SIZE], int n) {
    for (int i = 0; i < n; i++) {
        double suma = 0;
        for (int j = 0; j < i; j++)
            suma += L[i][j] * y[j];
        y[i] = (b[i] - suma) / L[i][i];
    }
}

// Lᵗx = y
void backward_sub(double L[MAX_SIZE][MAX_SIZE], double y[MAX_SIZE], double x[MAX_SIZE], int n) {
    for (int i = n - 1; i >= 0; i--) {
        double suma = 0;
        for (int j = i + 1; j < n; j++)
            suma += L[j][i] * x[j];
        x[i] = (y[i] - suma) / L[i][i];
    }
}

// Inversión con paralelismo en el ciclo por columnas
int invertir_cholesky(double A[MAX_SIZE][MAX_SIZE], double A_inv[MAX_SIZE][MAX_SIZE], int n) {
    double L[MAX_SIZE][MAX_SIZE] = {0};
    if (!cholesky(A, L, n)) return 0;

    #pragma omp parallel for
    for (int i = 0; i < n; i++) {
        double e[MAX_SIZE] = {0}, y[MAX_SIZE], x[MAX_SIZE];
        e[i] = 1.0;

        forward_sub(L, e, y, n);
        backward_sub(L, y, x, n);

        for (int j = 0; j < n; j++)
            A_inv[j][i] = x[j];
    }

    return 1;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Uso: %s <archivo_csv>\n", argv[0]);
        return 1;
    }

    char ruta_completa[1024];
    snprintf(ruta_completa, sizeof(ruta_completa), "%s%s", RUTA_BASE, argv[1]);

    double A[MAX_SIZE][MAX_SIZE], A_inv[MAX_SIZE][MAX_SIZE];
    int n;

    if (!leer_csv(ruta_completa, A, &n)) {
        return 1;
    }

    double inicio = omp_get_wtime();

    if (!invertir_cholesky(A, A_inv, n)) {
        printf("La matriz no es definida positiva o no es invertible.\n");
        return 1;
    }

    double fin = omp_get_wtime();
    double tiempo = fin - inicio;

    printf("Tiempo de inversión: %.6f segundos\n", tiempo);

    if (n <= 10) {
        printf("Matriz inversa:\n");
        imprimir(A_inv, n);
    } else {
        printf("Matriz mayor a 10x10. Inversa no mostrada en pantalla.\n");
    }

    return 0;
}

File: virtual/GR-06/codigo/cholesky/cholesky_paralelo.py
# -*- coding: utf-8 -*-
import numpy as np
import threading
import time
import sys
import os

def cargar_matriz_csv(ruta_csv):
    A = np.loadtxt(ruta_csv, delimiter=',')
    N = A.shape[0]
    print("Matriz A cargada desde CSV:")
    print(A)
    return A, N

def inicializar_matriz_inversa(N):
    matriz_inv = np.zeros((N, N))
    print("\nMatriz inversa inicializada en ceros:")
    print(matriz_inv)
    return matriz_inv

def descomposicion_cholesky(A):
    N = A.shape[0]
    #Se inicializa la matriz L
    L = np.zeros_like(A)
    #Recorre filas i y columnas j
    for i in range(N):
        for j in range(i+1):
            suma = 0.0
            #Elementos de la diagonal con la sumatoria de su formula para el calculo de L
            if j == i:
                for k in range(j):
                    suma += L[j,k] ** 2
                val = A[j,j] - suma
                if val <= 0:
                    raise ValueError(f"La matriz no es definida positiva en la posición ({j},{j})")
                L[j,j] = np.sqrt(val)
            else:
                for k in range(j):
                    suma += L[i,k] * L[j,k]
                L[i,j] = (A[i,j] - suma) / L[j,j]
    print("\nMatriz L (Cholesky) calculada:")
    print(L)
    return L

def forward_substitution(L, e):
    N = L.shape[0]
    y = np.zeros(N)
    for j in range(N):
        suma = 0.0
        for k in range(j):
            suma += L[j, k] * y[k]
        y[j] = (e[j] - suma) / L[j, j]
    return y

def backward_substitution(L, y):
    N = L.shape[0]
    x = np.zeros(N)
    for j in reversed(range(N)):
        suma = 0.0
        for k in range(j+1, N):
            suma += L[k, j] * x[k]
        x[j] = (y[j] - suma) / L[j, j]
    return x

def calcular_columna_inversa(i, L, matriz_inv):
    """Calcula la columna i-ésima de la matriz inversa y la guarda en matriz_inv."""
    N = L.shape[0]
    e = np.zeros(N)
    e[i] = 1
    y = forward_substitution(L, e)
    x = backward_substitution(L, y)
    matriz_inv[:, i] = x

def calcular_inversa_threading(L):
    N = L.shape[0]
    matriz_inv = np.zeros((N, N))
    threads = []

    for i in range(N):
        t = threading.Thread(target=calcular_columna_inversa, args=(i, L, matriz_inv))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    print("\nMatriz inversa calculada con Cholesky (threading):")
    print(matriz_inv)
    return matriz_inv

def main():
    if len(sys.argv) != 2:
        print("Uso: python3 cholesky_paralelo.py <dimension>")
        sys.exit(1)

    try:
        N = int(sys.argv[1])
        if N <= 0:
            raise ValueError
    except ValueError:
        print("La dimensión debe ser un número entero positivo.")
        sys.exit(1)

    ruta_csv = f'../../matrices/def_pos_inv/matriz_{N}x{N}_dpi.csv'

    if not os.path.isfile(ruta_csv):
        print(f"Archivo no encontrado: {ruta_csv}")
        sys.exit(1)

    # 1. Cargar matriz
    A, N = cargar_matriz_csv(ruta_csv)

    # 2. Inicializar matriz inversa (solo para mostrar)
    _ = inicializar_matriz_inversa(N)

    # 3. Calcular descomposición de Cholesky
    L = descomposicion_cholesky(A)

    # 4. Calcular inversa
    matriz_inv = calcular_inversa_threading(L)

if __name__ == "__main__":
    inicio = time.time()    # ⏱️ Marca de tiempo inicial
    main()
    fin = time.time()       # ⏱️ Marca de tiempo final
    print(f"\nTiempo de ejecución total: {fin - inicio:.4f} segundos")
